pkg/model/v2/segment_decoder.go

```diff
@@ -0,0 +1,115 @@
+package v2
+
+import (
+	"errors"
+	"fmt"
+	"math"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/grafana/tempo/pkg/model/trace"
+	"github.com/grafana/tempo/pkg/tempopb"
+)
+
+// SegmentDecoder maintains the relationship between distributor -> ingester
+// Segment format:
+//  | uint32 | uint32 | variable length          |
+//  | start  | end    | marshalled tempopb.Trace |
+// start and end are unix epoch seconds
+type SegmentDecoder struct {
+}
+
+var segmentDecoder = &SegmentDecoder{}
+
+func NewBatchDecoder() *SegmentDecoder {
+	return segmentDecoder
+}
+
+func (d *SegmentDecoder) PrepareForWrite(trace *tempopb.Trace, start uint32, end uint32) ([]byte, error) {
+	return marshalWithStartEnd(trace, start, end)
+}
+
+func (d *SegmentDecoder) PrepareForRead(batches [][]byte) (*tempopb.Trace, error) {
+	var combinedTrace *tempopb.Trace
+	for _, obj := range batches {
+		obj, _, _, err := stripStartEnd(obj)
+		if err != nil {
+			return nil, fmt.Errorf("error stripping start/end: %w", err)
+		}
+
+		t := &tempopb.Trace{}
+		err = proto.Unmarshal(obj, t)
+		if err != nil {
+			return nil, fmt.Errorf("error unmarshaling trace: %w", err)
+		}
+
+		combinedTrace, _ = trace.CombineTraceProtos(combinedTrace, t)
+	}
+
+	return combinedTrace, nil
+}
+
+// ToObject creates a byte slice that can be interpreted by ObjectDecoder in this package
+// see object_decoder.go for details on the format.
+func (d *SegmentDecoder) ToObject(batches [][]byte) ([]byte, error) {
+	// strip start/end from individual batches and place it in a TraceBytesWrapper
+	var err error
+	var minStart, maxEnd uint32
+	minStart = math.MaxUint32
+
+	for i, b := range batches {
+		var start, end uint32
+
+		batches[i], start, end, err = stripStartEnd(b)
+		if err != nil {
+			return nil, err
+		}
+		if start < minStart {
+			minStart = start
+		}
+		if end > maxEnd {
+			maxEnd = end
+		}
+	}
+
+	return marshalWithStartEnd(&tempopb.TraceBytes{
+		Traces: batches,
+	}, minStart, maxEnd)
+}
+
+func marshalWithStartEnd(pb proto.Message, start uint32, end uint32) ([]byte, error) {
+	const uint32Size = 4
+
+	sz := proto.Size(pb)
+	buff := make([]byte, 0, sz+uint32Size*2) // proto buff size + start/end uint32s
+
+	buffer := proto.NewBuffer(buff)
+
+	_ = buffer.EncodeFixed32(uint64(start)) // EncodeFixed32 can't return an error
+	_ = buffer.EncodeFixed32(uint64(end))
+	err := buffer.Marshal(pb)
```

Comment:
Is this marshal appending to the buffer?  So the two uint64s are put in place, and then the variable-length trace bytes?

Reply:
technically two uint32s, but yes. for some reason `EncodeFixed32` takes a uint64.

Reply:
Right sorry, `uint32`.